gitlab/integration_test.go

```diff
@@ -0,0 +1,674 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitlab
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gregjones/httpcache"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+)
+
+const (
+	ghTokenFile  = "/tmp/gitlab-token"
+	gitlabDomain = "gitlab.com"
+
+	defaultDescription = "Foo description"
+	// TODO: This will change
+	defaultBranch = masterBranchName
+	testUserName  = "dinosk"
+)
+
+var (
+	// customTransportImpl is a shared instance of a customTransport, allowing counting of cache hits.
+	customTransportImpl *customTransport
+)
+
+func init() {
+	// Call testing.Init() prior to tests.NewParams(), as otherwise -test.* will not be recognised. See also: https://golang.org/doc/go1.13#testing
+	testing.Init()
+	rand.Seed(time.Now().UnixNano())
+}
+
+func TestProvider(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "GitLab Provider Suite")
+}
+
+func customTransportFactory(transport http.RoundTripper) http.RoundTripper {
+	if customTransportImpl != nil {
+		panic("didn't expect this function to be called twice")
+	}
+	customTransportImpl = &customTransport{
+		transport:      transport,
+		countCacheHits: false,
+		cacheHits:      0,
+		mux:            &sync.Mutex{},
+	}
+	return customTransportImpl
+}
+
+type customTransport struct {
+	transport      http.RoundTripper
+	countCacheHits bool
+	cacheHits      int
+	mux            *sync.Mutex
+}
+
+func (t *customTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	resp, err := t.transport.RoundTrip(req)
+	// If we should count, count all cache hits whenever found
+	if t.countCacheHits {
+		if _, ok := resp.Header[httpcache.XFromCache]; ok {
+			t.cacheHits++
+		}
+	}
+	return resp, err
+}
+
+func (t *customTransport) resetCounter() {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	t.cacheHits = 0
+}
+
+func (t *customTransport) setCounter(state bool) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	t.countCacheHits = state
+}
+
+func (t *customTransport) getCacheHits() int {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	return t.cacheHits
+}
+
+func (t *customTransport) countCacheHitsForFunc(fn func()) int {
+	t.setCounter(true)
+	t.resetCounter()
+	fn()
+	t.setCounter(false)
+	return t.getCacheHits()
+}
+
+var _ = Describe("GitLab Provider", func() {
+	var (
+		ctx context.Context = context.Background()
+		c   gitprovider.Client
+
+		testSharedOrgRepoName string = "testsharedorgrepo"
+		testOrgRepoName       string = "testorgrepo"
+		testRepoName          string = "testrepo"
+		testOrgName           string = "GGPGroup"
+		testSubgroupName      string = "ggpsub"
+		testTeamName          string = "GGPGroup2"
```

Comment:
We should make a note here about which of these are expected to already exist vs. will be created

Reply:
Added comment.